### DISCUSSION TITLE
🚀 1.0.0-beta

### DISCUSSION THREAD

**Discussion by x-0o0** [x-0o0/ChatUI#3] [Announcements]
## What's New
* Chat in Channel & Appearance by @jaesung-0o0 in https://github.com/jaesung-0o0/ChatUI/pull/2

## Essentials
* To see the previews, see `/Sources/Previews`
* To see the developer documentation, see `/Documentation`
* To see the real use cases examples, go to [ChatUI-examples](https://github.com/jaesung-0o0/ChatUI-examples)

## New Contributors
* @jaesung-0o0 made their first contribution in https://github.com/jaesung-0o0/ChatUI/pull/2

**Full Changelog**: https://github.com/jaesung-0o0/ChatUI/commits/1.0.0-beta.0

<hr /><em>This discussion was created from the release <a href='https://github.com/jaesung-0o0/ChatUI/releases/tag/1.0.0-beta.0'>1.0.0-beta.0</a>.</em>

**Comment by x-0o0** [comment]
# NOTICE

Please let me know your issues freely!

**Comment by x-0o0** [comment]
# 🚀 1.0.0-beta.1 has been released

## What's Changed
* Publishers and modifiers regarding keyboard by @jaesung-0o0 in https://github.com/jaesung-0o0/ChatUI/pull/7
* Updated descriptions by @jaesung-0o0 in https://github.com/jaesung-0o0/ChatUI/pull/8

**Full Changelog**: https://github.com/jaesung-0o0/ChatUI/compare/1.0.0-beta.0...1.0.0-beta.1

**Comment by x-0o0** [comment]
# 🚀 1.0.0-beta.2 has been released

## What's Changed
* [Release/1.0.0-beta.2] Message Menu by @jaesung-0o0 in https://github.com/jaesung-0o0/ChatUI/pull/10


**Full Changelog**: https://github.com/jaesung-0o0/ChatUI/compare/1.0.0-beta.1...1.0.0-beta.2

## How to show message menu on long press gesture

You can add message menus to display when a `rowContent`(such as `MessageRow`) is on long press gesture by setting `menuContent` parameter of the `MessageList` initializer.

`MessageMenu` and `MessageMenubuttonStyle` allow you to create message menu more easily. Here is an example:

```swift
MessageList(messages) { message in
    // row content
    MessageRow(message: message)
        .padding(.top, 12)
} menuContent: { highlightMessage in 
    // menu content
    MessageMenu {
        Button("Copy", action: copy)
            .buttonStyle(MessageMenuButtonStyle(symbol: "doc.on.doc"))
        
        Divider()
            
        Button("Reply", action: reply)
            .buttonStyle(MessageMenuButtonStyle(symbol: "arrowshape.turn.up.right"))
            
        Divider()
        
        Button("Delete", action: delete)
            .buttonStyle(MessageMenuButtonStyle(symbol: "trash"))
    }
    .padding(.top, 12)
}
```

**Comment by x-0o0** [comment]
# 🚀 1.0.0-beta.3 has been released

## What's Changed
* Added `MessageDateView` by @jaesung-0o0 in https://github.com/jaesung-0o0/ChatUI/pull/12
* Modified `highlightMessagePublisher` to send optional value by @jaesung-0o0 in https://github.com/jaesung-0o0/ChatUI/pull/13


**Full Changelog**: https://github.com/jaesung-0o0/ChatUI/compare/1.0.0-beta.2...1.0.0-beta.3

**Comment by x-0o0** [comment]
# 🚀 1.0.0-beta.4 has been released

## What's Changed
* Bug-fix on `MessageDateView/date` by @jaesung-0o0 in https://github.com/jaesung-0o0/ChatUI/pull/14


**Full Changelog**: https://github.com/jaesung-0o0/ChatUI/compare/1.0.0-beta.3...1.0.0-beta.4

**Comment by x-0o0** [comment]
# 🚀 1.0.0-beta.5 has been released

## What's Changed
* [Release/1.0.0-beta.5] Fixed `highlightMessagePublisher` to send `nil` when tap blur background view by @jaesung-0o0 in https://github.com/jaesung-0o0/ChatUI/pull/17


**Full Changelog**: https://github.com/jaesung-0o0/ChatUI/compare/1.0.0-beta.4...1.0.0-beta.5

### Implementation Guide

```swift
MessageList(message)
    .onReceive(highlightMessagePublisher) { highlightMessage in
        withAnimation {
            self.highlightMessage = highlightMessage as? Message
        }
    }
            
if highlightMessage == nil {
    MessageField { messageStyle in
        print(messageStyle)
    }
}
```

**Comment by x-0o0** [comment]
# 🚀 1.0.0-beta.6 has been released

## What's Changed
* [ADD] Added `MessageRow/lineLimit` and `MessageField/characterLimit` by @jaesung-0o0 in https://github.com/jaesung-0o0/ChatUI/pull/21


**Full Changelog**: https://github.com/jaesung-0o0/ChatUI/compare/1.0.0-beta.5...1.0.0-beta.6

Supports line limit in `MessageRow` and character limit in `MessageField`. Now both have an unlimited line(or character) as a default.

Here is an example usage:

```swift
MessageList(messages) { message
    // rowContent
    MessageRow(message: message, lineLimit: 10) // 🌟 set `lineLimit`. The default is `nil` which means unlimited
        .padding(.top, 12)
}

MessageField(characterLimit: 300) { messageStyle in // 🌟 set `characterLimit`. The default is `nil` which means unlimited
    // send message...
}
```

### Additional Information

💡 It is not necessary to use `MessageRow` in the `rowContent` of `MessageList`. It's able to design your own message row that suits your app's needs by using the **`messageStyle(_:)` modifier in `View`** that is provided by ChatUI.

- Example usage:

```swift
MessageList(messages) { message in
    HStack(alignment: .top, spacing: 8) {
        AsyncImage(url: message.sender.imageURL) { image in
            image
                .resizable()
                .messageStyle(.senderProfile)
            } placeholder: {
                Image(systemName: "person.crop.circle.fill")
                    .resizable()
                    .messageStyle(.senderProfile)
                    .foregroundColor(appearance.secondary)
            }

        VStack(alignment: .leading, spacing: 8) {
            switch style {
            case .text(let text):
                let markdown = LocalizedStringKey(text)
                Text(markdown)
                    .fixedSize(horizontal: false, vertical: true) // or use `.lineLimit(nil`
                    .tint(isMyMessage ? appearance.prominentLink : appearance.link)
                    .messageStyle(isMyMessage ? .localBody({LINE.LIMIT}) : .remoteBody({LINE.LIMIT}))
            default:
                Text("Not supported format")
            }

            Text(Date(timeIntervalSince1970: message.sentAt), formatter: formatter)
                .messageStyle(.date)
        }
    }
}
```

**Comment by x-0o0** [comment]
# 🚀 1.0.0-beta.7 has been released

## What's Changed
* Renamed `Appearance/messagebody` to `Appearance/body` by @jaesung-0o0 in https://github.com/jaesung-0o0/ChatUI/pull/24
* [ADD] Added `public init` to message views by @jaesung-0o0 in https://github.com/jaesung-0o0/ChatUI/pull/23
* [FIX] Allowed nullable `MessageList/menuContent` by @jaesung-0o0 in https://github.com/jaesung-0o0/ChatUI/pull/22
* [ADD] Added menu content condition to handle `highlightMessagePublisher` by @jaesung-0o0 in https://github.com/jaesung-

**Full Changelog**: https://github.com/jaesung-0o0/ChatUI/compare/1.0.0-beta.6...1.0.0-beta.7

**Comment by x-0o0** [comment]
# 🚀 1.0.0-beta.8 has been released

## What's Changed
* [FIX] Fixed issue that the list isn't scrolled when tap message row by @jaesung-0o0 in https://github.com/jaesung-0o0/ChatUI/pull/28


**Full Changelog**: https://github.com/jaesung-0o0/ChatUI/compare/1.0.0-beta.7...1.0.0-beta.8

**Comment by x-0o0** [comment]
# 🚀 1.0.0-beta.9 has been released

## What's Changed
* Updated to allow configuration of Giphy UI by @fred-bowker in https://github.com/jaesung-0o0/ChatUI/pull/31
* [FIX] Adjusted code styles by @jaesung-0o0 in https://github.com/jaesung-0o0/ChatUI/pull/32
* Added `NextMessageField` to provide easy-customizable contents by @jaesung-0o0 in https://github.com/jaesung-0o0/ChatUI/pull/33
* Added `scrolledToEndPublisher` (Beta) by @jaesung-0o0 in https://github.com/jaesung-0o0/ChatUI/pull/34

## New Contributors
* @fred-bowker made their first contribution in https://github.com/jaesung-0o0/ChatUI/pull/31

**Full Changelog**: https://github.com/jaesung-0o0/ChatUI/compare/1.0.0-beta.8...1.0.0-beta.9

## Example Usage - NextMessageField

```swift
 @State private var text: String = ""
 
 NextMessageField(text) { messageStyle in
    guard !text.isEmpty else { return }
    viewModel.sendMessage($0)
    text = ""
 } leftLabel: {
    HStack {
        Button(aciton: showCamera) {
            Image.camera.medium
        }
        .frame(width: 36, height: 36)
 
        Button(action: showLibrary) {
            Image.photoLibrary.medium
        }
        .frame(width: 36, height: 36)
 } rightLabel: {
    Button {
        // send message by using `sendMessagePublisher`. This will invoke `onSend` handler.
        sendMessagePublisher.send(.text(text))
    } label: {
        // send button icon
        Image.send.medium
    }
    .frame(width: 36, height: 36)
 }
 ```

## Example Usage - GiphyConfiguration

```swift
let giphyConfig = GiphyConfiguration(
    dimBackground: true, 
    presentationDetents = 0.7
)
let chatConfig = ChatConfiguration(
    userID: "{USER.ID}", 
    giphyKey: "{YOUR.GIPHY.KEY}", 
    giphyConfig: giphyConfig
)

// ...

ChatView()
    .environmentObject(chatConfig)
```

## Notice

`scrolledToEndPublisher` publisher is added, but because it's a beta feature, I'm not responsible for any issues that may arise during using it.

**Comment by fred-bowker** [comment]
Cheers for your work on this, it's a great framework, thanks for getting the release out :)

**Reply by x-0o0** [reply]
THANK YOU!! 👍
